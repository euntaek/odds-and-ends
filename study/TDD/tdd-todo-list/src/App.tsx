import styled from "styled-components";
import { PageHeader } from "Components";
import { List, Add } from "Pages";

function App() {
  return (
    <Container>
      <PageHeader />
      <Add />
    </Container>
  );
}

const Container = styled.div`
  min-height: 100vh;
  background-color: #eeeeee;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

export default App;
