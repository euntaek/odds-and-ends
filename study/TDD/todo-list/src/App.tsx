import { useState } from "react";
import styled from "styled-components";

import { Button, Input, ToDoItem } from "Components";

function App() {
  const [toDo, setToDo] = useState("");
  const [toDoList, setToDoList] = useState<string[]>([]);

  const addToDo = () => {
    if (toDo) {
      setToDoList((toDoList) => [...toDoList, toDo]);
      setToDo("");
    }
  };

  const deleteToDo = (index: number) => {
    setToDoList((toDoList) => toDoList.filter((toDo, idx) => idx !== index));
  };

  return (
    <Container>
      <Contents>
        <ToDoListContainer data-testid="toDoList">
          {toDoList.map((toDo, index) => (
            <ToDoItem key={toDo} label={toDo} onDelete={() => deleteToDo(index)} />
          ))}
        </ToDoListContainer>
        <InputContainer>
          <Input
            placeholder="할 일을 입력해 주세요"
            value={toDo}
            onChange={(text) => {
              setToDo(text);
            }}
          />
          <Button label="추가" onClick={addToDo} />
        </InputContainer>
      </Contents>
    </Container>
  );
}

const Container = styled.div`
  min-height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

const Contents = styled.div`
  display: flex;
  background-color: #ffffff;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;

const InputContainer = styled.div`
  display: flex;
`;

const ToDoListContainer = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
`;

export default App;
