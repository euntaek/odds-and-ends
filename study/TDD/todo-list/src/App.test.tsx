import React from "react";
import { render, fireEvent } from "@testing-library/react";

import App from "./App";

describe("<App/>", () => {
  test("renders component correctly", () => {
    const { container, getByTestId, getByText, getByPlaceholderText } = render(<App />);

    const toDoList = getByTestId("toDoList");
    expect(toDoList).toBeInTheDocument();
    expect(toDoList.firstChild).toBeNull();

    const input = getByPlaceholderText("할 일을 입력해 주세요");
    expect(input).toBeInTheDocument();
    const label = getByText("추가");
    expect(label).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("adds and deletes toDo items", () => {
    const { getByText, getAllByText, getByPlaceholderText, getByTestId } = render(<App />);
    const input = getByPlaceholderText("할 일을 입력해 주세요");
    const button = getByText("추가");
    fireEvent.change(input, { target: { value: "study react 1" } });
    fireEvent.click(button);

    const toDoItem = getByText("study react 1");
    expect(toDoItem).toBeInTheDocument();
    const deletebutton = getByText("삭제");
    expect(deletebutton).toBeInTheDocument();

    const toDoList = getByTestId("toDoList");
    expect(toDoList.childElementCount).toBe(1);

    fireEvent.change(input, { target: { value: "study react 2" } });
    fireEvent.click(button);

    const toDoItem2 = getByText("study react 2");
    expect(toDoItem2).toBeInTheDocument();
    expect(toDoList.childElementCount).toBe(2);

    const deleteButtons = getAllByText("삭제");
    expect(deleteButtons).toHaveLength(2);

    fireEvent.click(deleteButtons[0]);
    expect(toDoItem).not.toBeInTheDocument();
    expect(toDoList.childElementCount).toBe(1);
  });

  test("does not add empty ToDo", () => {
    const { getByTestId, getByText } = render(<App />);

    const toDoList = getByTestId("toDoList");
    expect(toDoList).toBeInTheDocument();
    const button = getByText("추가");
    expect(button).toBeInTheDocument();

    const length = toDoList.childElementCount;
    fireEvent.click(button);

    expect(toDoList.childElementCount).toBe(length);
  });
});
