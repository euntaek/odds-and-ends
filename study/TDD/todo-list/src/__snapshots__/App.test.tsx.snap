// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders component correctly 1`] = `
<div>
  <div
    class="sc-jSFjdj geTXlp"
  >
    <div
      class="sc-gKAaRy fsYFLu"
    >
      <div
        class="sc-fujyAs grBJRJ"
        data-testid="toDoList"
      />
      <div
        class="sc-iCoGMd icFpiP"
      >
        <input
          class="sc-dlnjwi eaATVp"
          placeholder="할 일을 입력해 주세요"
          value=""
        />
        <button
          class="sc-bdnxRM dnLCMG"
        >
          <div
            class="sc-gtsrHT fEppky"
          >
            추가
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
